File: packages/i18nify-js/src/modules/dateTime/getFinancialYear.ts
import { withErrorBoundary } from '../../common/errorBoundary';
import { DateInput } from './types';
import { convertToStandardDate } from './utils';

export type FYLabelFormat = 'short' | 'long' | 'fy';

export interface GetFinancialYearOptions {
  labelFormat?: FYLabelFormat;
}

interface FYConfig {
  startMonth: number;
  labelFormat: FYLabelFormat;
}

// FY start month (1-indexed) and default label format per country.
const FY_CONFIG: Record<string, FYConfig> = {
  // April 1 → March 31  → "2024-25"
  IN: { startMonth: 4, labelFormat: 'short' },
  JP: { startMonth: 4, labelFormat: 'short' },
  GB: { startMonth: 4, labelFormat: 'short' },
  CA: { startMonth: 4, labelFormat: 'short' },
  NZ: { startMonth: 4, labelFormat: 'short' },
  SG: { startMonth: 4, labelFormat: 'short' },
  ZA: { startMonth: 4, labelFormat: 'short' },
  // July 1 → June 30  → "2024-25"
  AU: { startMonth: 7, labelFormat: 'short' },
  BD: { startMonth: 7, labelFormat: 'short' },
  PK: { startMonth: 7, labelFormat: 'short' },
  // October 1 → September 30  → "FY2025"
  US: { startMonth: 10, labelFormat: 'fy' },
};

const SUPPORTED_COUNTRIES = Object.keys(FY_CONFIG).join(', ');

const buildLabel = (
  fyStartYear: number,
  fyEndYear: number,
  format: FYLabelFormat,
): string => {
  switch (format) {
    case 'short':
      return `${fyStartYear}-${String(fyEndYear).slice(-2)}`;
    case 'long':
      return `${fyStartYear}-${fyEndYear}`;
    case 'fy':
      return `FY${fyEndYear}`;
  }
};

const getFinancialYear = (
  date: DateInput,
  countryCode: string,
  options: GetFinancialYearOptions = {},
): string => {
  if (date === null || date === undefined || date === '')
    throw new Error(
      `Parameter 'date' is invalid! The received value was: ${date}.`,
    );
  if (!countryCode)
    throw new Error(
      `Parameter 'countryCode' is invalid! The received value was: ${countryCode}.`,
    );

  const config = FY_CONFIG[countryCode.toUpperCase()];
  if (!config)
    throw new Error(
      `Country code "${countryCode}" is not supported for financial year calculation. Supported countries: ${SUPPORTED_COUNTRIES}.`,
    );

  const d = convertToStandardDate(date);
  const year = d.getFullYear();
  const month = d.getMonth() + 1; // convert to 1-indexed

  const fyStartYear = month >= config.startMonth ? year : year - 1;
  const fyEndYear = fyStartYear + 1;
  const labelFormat = options.labelFormat ?? config.labelFormat;

  return buildLabel(fyStartYear, fyEndYear, labelFormat);
};

export default withErrorBoundary<typeof getFinancialYear>(getFinancialYear);

File: packages/i18nify-go/modules/datetime/get_financial_year.go
package datetime

import (
	"fmt"
	"strings"
	"time"
)

// FYLabelFormat controls the financial year label style.
type FYLabelFormat string

const (
	FYLabelShort FYLabelFormat = "short"
	FYLabelLong  FYLabelFormat = "long"
	FYLabelFY    FYLabelFormat = "fy"
)

// GetFinancialYearOptions holds optional overrides for GetFinancialYear.
type GetFinancialYearOptions struct {
	LabelFormat FYLabelFormat
}

type fyCountryConfig struct {
	startMonth  int
	labelFormat FYLabelFormat
}

var (
	fySupportedCountries = []string{"IN", "JP", "GB", "CA", "NZ", "SG", "ZA", "AU", "BD", "PK", "US"}
	fyConfigMap          = map[string]fyCountryConfig{
		"IN": {startMonth: 4, labelFormat: FYLabelShort},
		"JP": {startMonth: 4, labelFormat: FYLabelShort},
		"GB": {startMonth: 4, labelFormat: FYLabelShort},
		"CA": {startMonth: 4, labelFormat: FYLabelShort},
		"NZ": {startMonth: 4, labelFormat: FYLabelShort},
		"SG": {startMonth: 4, labelFormat: FYLabelShort},
		"ZA": {startMonth: 4, labelFormat: FYLabelShort},
		"AU": {startMonth: 7, labelFormat: FYLabelShort},
		"BD": {startMonth: 7, labelFormat: FYLabelShort},
		"PK": {startMonth: 7, labelFormat: FYLabelShort},
		"US": {startMonth: 10, labelFormat: FYLabelFY},
	}
)

// GetFinancialYear returns the financial year label for a date and country code.
func GetFinancialYear(date time.Time, countryCode string, opts ...GetFinancialYearOptions) (string, error) {
	code := strings.ToUpper(strings.TrimSpace(countryCode))
	if code == "" {
		return "", fmt.Errorf("getFinancialYear: country code must not be empty")
	}

	config, exists := fyConfigMap[code]
	if !exists {
		return "", fmt.Errorf(
			"getFinancialYear: country code %q is not supported; supported: %s",
			countryCode,
			strings.Join(fySupportedCountries, ", "),
		)
	}

	labelFormat := config.labelFormat
	if len(opts) > 0 && opts[0].LabelFormat != "" {
		labelFormat = opts[0].LabelFormat
	}

	fyStartYear := date.Year()
	if int(date.Month()) < config.startMonth {
		fyStartYear--
	}

	fyEndYear := fyStartYear + 1

	switch labelFormat {
	case FYLabelShort:
		return fmt.Sprintf("%d-%02d", fyStartYear, fyEndYear%100), nil
	case FYLabelLong:
		return fmt.Sprintf("%d-%d", fyStartYear, fyEndYear), nil
	case FYLabelFY:
		return fmt.Sprintf("FY%d", fyEndYear), nil
	default:
		return "", fmt.Errorf("getFinancialYear: unknown label format %q", labelFormat)
	}
}
